core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -296,6 +300,89 @@ def validate_leading_factor_of_delay(
+    fund_type = workbook["Programme_Ref"]["FundType_ID"][0]
+    project_ids = workbook["Project Details"]["Project ID"]
+    funding_df = workbook["Funding"]
+
+    # pull funding spent for individual projects and programme wide total
+    funding_spent = {project: spend_per_project(funding_df, project) for project in project_ids}
+    funding_spent[programme_id] = sum(funding_spent.values())
+
+    funding_allocated = get_allocated_funding()
+
+    funding_spent_failures = []
+    if fund_type == "HS":
+        # check funding against programme wide funding allocated for Future High Street Fund submissions
+        validate_by = "programme"
+        ids_to_check = [programme_id]
+
+    else:
+        # check funding against individual project funding allocated for Towns Deal submissions
+        validate_by = "project"
+        ids_to_check = project_ids.tolist()
+
+    for idx in ids_to_check:
+        if funding_spent[idx] > funding_allocated[idx]:
+            funding_spent_failures.append(
+                TownsFundRoundFourValidationFailure(
+                    sheet="Funding Profiles",
+                    section="Project Funding Profiles"
+                    + (f" - Project {get_project_number(idx)}" if validate_by == "project" else ""),
+                    column="Grand Total",
+                    message=(
+                        f"The total spend for this {validate_by} is higher than amount allocated for the {validate_by}."
+                        f" Please check the total spend and resubmit your spreadsheet."
+                        f" You spent {funding_spent[idx]} but were only allocated {funding_allocated[idx]}"
+                    ),
+                    row_indexes=[15 + 28 * get_project_number(idx)]
+                    if validate_by == "project"
+                    else [15 + 28 * get_project_number(proj_id) for proj_id in project_ids],
+                )
+            )
+
+    return funding_spent_failures
+
+
+def spend_per_project(funding_df: pd.DataFrame, project_id: str) -> float:
+    """return the total funding spent per an individual project
+
+    :param funding_df: A dataframe of the funding table from a submission
+    :param: project_id: ID of project in question
+
+    :return: Total funding spent per project
+    """
+    funding_spent = (
+        funding_df["Spend for Reporting Period"]
+        .loc[
+            (funding_df["Project ID"] == project_id)
+            & (funding_df["Funding Source Type"] == "Towns Fund")
+            & ~(funding_df["Funding Source Name"].str.contains(r"contractually committed"))
+        ]
+        .sum()
+    )
+    # Business logic here is taken from spreadsheet 4a - Funding Profile Z45 for grand total expenditure
+    return funding_spent
+
+
+def get_allocated_funding():
+    return pd.read_csv(
+        get_file(DefaultConfig.AWS_S3_BUCKET_FILE_ASSETS, TF_FUNDING_ALLOCATED_FILE), index_col="Index Code"
+    )["Grant Awarded"]
```

Comment:
My issue with this was that it's then done every request. This almost certainly won't make a practical difference at our level of scale. It just seems like better practice to load this into memory once per application initialisation and then fetch from there instead of repeatedly doing this I/O. I feel like there has to be clean way to implement this.. 

What do you think @Connor-OS @nsnape?

Reply:
Could we just instantiate it as a constant at the top of this module from this function and use that within `validate_funding_spent`? The way we mock wouldn't have to change

Reply:
I I think the way I spoke about it with Noel earlier was that it can't be a constant if it is being dynamically read, it is more of a config prehaps, but that has other issues.

When do you think it will be reused? When users reupload after fixing errors? I guess that will only be 1 or 2 times on average.

Reply:
This way we will be reading the file from S3 every single request that is made. The reuse would be the lifetime of the application - so a fair bit. Perhaps constant isn't the correct term, I just mean a variable that lasts the lifetime of the application, it could be stored in this module for instance.

Reply:
We've looked into doing this as it would be more effecient to load this data just once on application initialisation. However it has made things to complex to mock this function call. We've deprioritised the issue for now and will add a ticket to the backlog to be picked up after delivery